### DISCUSSION TITLE
プログラムの国際化を検討したい

### DISCUSSION THREAD

**Discussion by berryzplus** [sakura-editor/sakura#1782] [Ideas]
## 議題
サクラエディタの国際化を行いたいです。

## ディスカッションの目的
国際化を行うにあたって、考えないといけないことの認識共有を行うことが目的です。

## 大前提
サクラエディタは日本語アプリです。
日本語環境下ならメニューやダイアログを英語表示できますが、日本語アプリです。
英語環境下で動かすと文書が文字化けします。

## 国際化とは？
国際化とは、いかなる言語環境下であっても動作できるようにプログラムを再設計することです。

サクラエディタは v2 で内部処理の大半を UNICODE化 しているため、プログラムの多くの部分はそのまま使えます。
しかし、独自の文字コード変換処理などに SJIS想定 のものが残っていたり、グローバル領域にローカライズが必要な文字列を保持していたりで再設計が必要な部分が残っています。

国際化アプリを作るために、何を考えたらいいかのチェックリストは MS のドキュメントとしても公開されています。
https://docs.microsoft.com/ja-jp/windows/win32/intl/internationalization-checklist

現状は、英語環境下で文字化けが起きるので「国際化できていない」の認識です。

## この議論の終了条件
以下いずれかが満たされたら閉じます。
1. サクラエディタの国際化が完了した。
2. 「国際化未達」が誤解であると判明した。
3. サクラエディタは国際化しないのプロジェクト方針が合意された。

議論と言っても議論しようがないかもしれないですが・・・。

**Comment by beru** [comment]
> 英語環境下で動かすと文書が文字化けします。

英語環境下というか外国語版のWindowsで日本語をインストールしていない動作環境ですかね？

文字化けするのはフォントが動作環境に無いからですかね？もしそうであればフォントの存在チェックとかフォント作成が失敗した時のエラー処理というか代替処理が不足しているのかなと思います。

**Reply by k-takata** [reply]
> 英語環境下で動かすと文書が文字化けします。

これがどういう条件なのかは分かりませんが、少なくとも、

> 独自の文字コード変換処理などに SJIS想定 のものが残っていたり、

については #1578 で直ったのではないかと勝手に思っています。

**Reply by berryzplus** [reply]
> > 英語環境下で動かすと文書が文字化けします。
> 
> 英語環境下というか外国語版のWindowsで日本語をインストールしていない動作環境ですかね？

再現条件は「システムロケールが "ja-JP" 以外」だと思っています。
Windows設定の「Unicode対応ではないプログラムの言語」を「日本語(日本)」以外にしたときに、
UTF-8やEUC-JPのファイルを開くと発生するようです。

SJISに関しては直ってるのかもしれないです。

**Reply by beru** [reply]
確かに正しい再現条件はそれのような気がしてきました。
それならば確認する環境を比較的用意しやすいので、動作確認して本当に問題が起きるか調べてみるのが良さそうですね。

**Reply by k-takata** [reply]
現時点で最新のビルド https://github.com/sakura-editor/sakura/actions/runs/2851309724 をダウンロードして、[MSEdge評価用仮想マシン](https://developer.microsoft.com/ja-jp/microsoft-edge/tools/vms/)上で動かしてみましたが、UTF-8/SJIS/EUC-JP いずれも文字化けせずにファイルを読み書きできるのを確認しました。#1578 で直っています。
（v2.4.1正式版では想定通り文字化けしました。）

**Reply by berryzplus** [reply]
いったん「英語環境下で動かすと文書が文字化けします。」は「誤認」で確認とれたと思ってます。
文字化けを解消するためのTODOは「最新ソースをリリースする」だと思います。

まだ残件あるので閉じられない認識です。

**Comment by k-takata** [comment]
英語版でメニューが `File(F)` のように表示されるのは違和感があります。
Notepad は `File` のようになっており、`(F)` のような括弧書きはありません。
![image](https://user-images.githubusercontent.com/840186/189479259-fb87907e-e9f6-435a-8d66-617d576b9def.png)

**Reply by k-takata** [reply]
あと、sakuraのウィンドウをフォアグラウンドにしたところ、ファイルのパスが英語版なのに `デスクトップ\...` と表示されたのは奇妙ですね。
![image](https://user-images.githubusercontent.com/840186/189479495-3ccc835d-1315-4dc5-9052-03a4852318e2.png)
